### DISCUSSION TITLE
Method to add file to File Manager

### DISCUSSION THREAD

**Discussion by blairliikala** [ExpressionEngine/ExpressionEngine#2873] [Feature Requests]
Add a simpler method to add files to the File Manager from remote locations.  Useful for a number of purposes, and many file API's offer adding files via a URL.

Here's an example.  You have a file on another server or system, and want to add it to EE's File Manager in a programatic way.  The steps are first to copy the file locally by getting path locations from the destination model, then copying the file, then adding the file through the file model.  It is a lot of steps that require the developer to navigate a few models for a fairly routine operation.

What adding a file should look like is:
```php
$url = 'http://example.com/myfile';
$overrides = [
   'name' => 'Custom name example.',
   'destination_id' => 1 // specify the destination, or fallback to the default one.
];
$model = ee('FileManager')->add($url, $overrides);

// Ideally, but might be tricky:
$model->status; // 'transferring' or 'downloading'
```

**Comment by intoeetive** [comment]
@blairliikala are you willing to actually fetch & copy over the file and have it added to EE, on this is something in Cloud Storage container that you want to make available for File Manager?

In the latter case, you could just use `ee('Model')->make('File')` etc (with the storage location being assigned as Upload Directorty)

**Comment by blairliikala** [comment]
In this case the file would be already at a location the web server running EE could access, but not part of a location being managed by the File Manager.  So it could be an S3 bucket or GCP bucket, but maybe from a different owner.  It could be another web server, or dropbox link.  It could be a file already on the EE web server, placed there by a process or script outside of EE.

Yes, it is possible to use the File model to do this, but it takes a bit of code and knowledge to make it work.  The request here is to make it simpler to add files to EE, and decrease developer time on it.  For me, it was surprising this wasn't already an option.

For example, the developer would first need to know that File and Upload Destinations are different models, but both are needed to usually get a file added.  File just creates the database entry, not actually adds the file itself which can be confusing.  If a destination ID isn't specified, the developer needs to know the useful default ID is actually 4.  While the file model only needs the destination ID, to actually copy the file you need the destination path.  Also handling is needed if the file name already exists, or using a UID filename approach.

**Comment by intoeetive** [comment]
@blairliikala with all the functionality you describe summed up, this looks more like an add-on. But - adding this to FR list
